### PR TITLE
[FLINK-14811][runtime] Replace Java Streams in input check methods with for-loops

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

Vertex input checking is invoked in lazily triggered scheduling by a FINISHED vertex state update RPC or a scheduleOrUpdateConsumers RPC.
Java Streams is used in it, but it should be avoided since it is performance critical code. 
This PR is to refactor these Java Streams to improve the performance, for both legacy scheduler and NG schedulers.

## Brief change log

  - *Replaced Java Streams in input check methods with for-loops*

## Verifying this change

This change is already covered by existing tests, such as InputDependencyConstraintCheckerTest and ExecutionVertexInputConstraintTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java [hunk 1]
```diff
@@ -106,7 +106,6 @@ public void testInputConstraintANY() throws Exception {
 		eg.start(mainThreadExecutor);
 		eg.scheduleForExecution();
 
-
```
Comment (TisonKun) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Unrelated changes?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java [hunk 2]
```diff
@@ -142,7 +142,7 @@ private void allocateSlotsAndDeployExecutionVertices(
 
 		schedulerOperations.allocateSlotsAndDeploy(
 			IterableUtils.toStream(vertices)
-				.filter(isInputConstraintSatisfied().and(IS_IN_CREATED_EXECUTION_STATE))
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java]: It seems we still use Java Stream here

Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java]: ok I notice that it is in a separated commit for an on-the-fly improvement. Make sense.

**Comment by zhuzhurk** [comment]
> Thanks for opening this @zhuzhurk ! The purpose sounds reasonable to me. Is there a benchmark report for this improvement?

It's about 2x speed up if the job scale is large.  see [this discussion](https://issues.apache.org/jira/browse/FLINK-14735?focusedCommentId=16974209&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16974209).
It's also suggested in the [flink code style guide](https://flink.apache.org/contributing/code-style-and-quality-java.html).

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 1 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: That's right. Let me move it into a hotfix.

Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: done.

Comment (TisonKun) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Interesting, why not this checkstyle violations complain before?

Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: Some flink-runtime classes style checks are suppressed for now because there were lots of violations when the checkstyle was introduced.
It should be good to fix them by the way, so that one day in the future we can remove those checkstyle suppression.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 3]
```diff
@@ -804,13 +802,37 @@ void cachePartitionInfo(PartitionInfo partitionInfo){
...
+			case ANY:
+				return isAnyInputConsumable();
+			case ALL:
+				return areAllInputsConsumable();
+			default:
+				throw new IllegalStateException("Unknown InputDependencyConstraint " + inputDependencyConstraint);
+		}
+	}
+
+	boolean isAnyInputConsumable() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: private?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 4]
```diff
@@ -804,13 +802,37 @@ void cachePartitionInfo(PartitionInfo partitionInfo){
...
+		for (int inputNumber = 0; inputNumber < inputEdges.length; inputNumber++) {
+			if (isInputConsumable(inputNumber)) {
+				return true;
+			}
 		}
+		return false;
+	}
+
+	boolean areAllInputsConsumable() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: private?

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionVertexInputConstraintTest.java]: we only wanted to make these changes though when we touch a file for other changes though; the only reason why we aren't applying checkstyle everywhere is the potential for merge conflicts; if we now go around and randomly apply checkstyle we still have the same risk.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 4 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: yes. done.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 3 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: done.

**Comment by zhuzhurk** [comment]
Rebased onto latest master.

**Comment by zentol** [comment]
Will merge this excluding the checkstyle changes as they are covered in #10278

**Comment by zhuzhurk** [comment]
Thanks a lot for merging the changes! @zentol
